### PR TITLE
prov/verbs: Enabling MR caching

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
This PR bases on the #3548 PR's changes
The intent of this PR is to add support of MR manager/MR cache for the verbs provider

### CODE REVIEW

**Comment by shefty** [comment]
How does this PR relate to #3548 ?

**Comment by shefty** [comment]
I'm assuming that the first 2 commits are the same ones from #3548 ?

File: prov/verbs/src/verbs_mr.c [hunk 1]
```diff
@@ -88,79 +164,138 @@ fi_ibv_mr_reg(struct fid *fid, const void *buf, size_t len,
...
-			fi_eq_write(&md->domain->util_domain.eq->eq_fid,
-				    FI_MR_COMPLETE, &entry, sizeof(entry), 0);
-	}
-	return 0;
-
+	return FI_SUCCESS;
 err:
 	free(md);
-	return -errno;
+	return -ret;
```
Comment (a-ilango) [inline prov/verbs/src/verbs_mr.c]: This should just be ret. fi_ibv_mr_reg_common already negates errno when it returns.

**Comment by gladkovdmitry17** [comment]
@shefty yes, you’re right. I’m going to exclude them, when the #3548 will be merged

File: prov/verbs/src/verbs_mr.c [hunk 2]
```diff
@@ -88,79 +164,138 @@ fi_ibv_mr_reg(struct fid *fid, const void *buf, size_t len,
...
+	if (OFI_UNLIKELY(flags))
+		return -FI_EBADFLAGS;
+	if (OFI_UNLIKELY(fid->fclass != FI_CLASS_DOMAIN))
+		return -FI_EINVAL;
+	if (OFI_UNLIKELY(attr->iov_count > VERBS_MR_IOV_LIMIT)) {
 		VERBS_WARN(FI_LOG_FABRIC,
 			   "iov count > %d not supported\n",
 			   VERBS_MR_IOV_LIMIT);
 		return -FI_EINVAL;
 	}
```
Comment (a-ilango) [inline prov/verbs/src/verbs_mr.c]: The checks above seem to be common for both this function and fi_ibv_mr_regattr. They can be moved into a check function to be called from both.

File: prov/verbs/src/verbs_mr.c [hunk 3]
```diff
@@ -33,6 +33,53 @@
...
+	};										\
+	const struct fi_mr_attr attr = {						\
+		.mr_iov		= &iov,							\
+		.iov_count	= 1,							\
+		.access		= access,						\
+		.offset		= offset,						\
+		.requested_key	= requested_key,					\
+		.context	= context,						\
+	};										\
+	return fi_ibv_mr ## type ## regattr(fid, &attr, flags, mr);			\
```
Comment (a-ilango) [inline prov/verbs/src/verbs_mr.c]: How about calling fi_ibv_mr_ ## regv (...) here?

File: prov/verbs/src/verbs_mr.c (hunk 1 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_mr.c]: Good catch, thanks!
Done

File: prov/verbs/src/verbs_mr.c (hunk 2 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_mr.c]: Done

File: prov/verbs/src/verbs_mr.c (hunk 3 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_mr.c]: Done

**Comment by gladkovdmitry17** [comment]
@shefty @a-ilango Please, welcome for the review.

I've rebased this patch on top of #3617 that has been merged yesterday

File: prov/verbs/src/verbs_mr.c [hunk 4]
```diff
@@ -130,37 +154,292 @@ fi_ibv_mr_reg(struct fid *fid, const void *buf, size_t len,
...
+
+void fi_ibv_monitor_unsubscribe(struct ofi_mem_monitor *notifier, void *addr,
+				size_t len, struct ofi_subscription *subscription)
+{
+}
+
+struct ofi_subscription *fi_ibv_monitor_get_event(struct ofi_mem_monitor *notifier)
+{
+	return NULL;
+}
```
Comment (shefty) [inline prov/verbs/src/verbs_mr.c]: These are fine for now as placeholders, but I think the final solution (guessing it will try to hook alloc/free) should put the monitor implementation into util code.

File: prov/verbs/src/verbs_domain.c [hunk 5]
```diff
@@ -300,7 +323,37 @@ fi_ibv_domain(struct fid_fabric *fabric, struct fi_info *info,
...
+		ofi_monitor_init(&_domain->monitor);
+
+		_domain->cache.size = fi_ibv_gl_data.mr_cache_size;
+		_domain->cache.entry_data_size = sizeof(struct fi_ibv_mem_desc);
+		_domain->cache.add_region = fi_ibv_mr_cache_entry_reg;
+		_domain->cache.delete_region = fi_ibv_mr_cache_entry_dereg;
+		ret = ofi_mr_cache_init(&_domain->util_domain, &_domain->monitor,
+					&_domain->cache);
+		if (ret)
+			goto err4;
```
Comment (a-ilango) [inline prov/verbs/src/verbs_domain.c]: ofi_monitor_init just initializes an atomic variable. If there isn't a future plan to extend the function to do something more, going to err4 is probably fine here, otherwise we may need to call ofi_monitor_cleanup before that.

File: prov/verbs/src/verbs_mr.c [hunk 6]
```diff
@@ -130,37 +154,292 @@ fi_ibv_mr_reg(struct fid *fid, const void *buf, size_t len,
...
+		ibv_access |= IBV_ACCESS_REMOTE_READ;
+
+	/* Verbs requires Local Write access too for Remote Write access */
+	if (ofi_access & FI_REMOTE_WRITE)
+		ibv_access |= IBV_ACCESS_LOCAL_WRITE |
+			      IBV_ACCESS_REMOTE_WRITE |
+			      IBV_ACCESS_REMOTE_ATOMIC;
+
+	return ibv_access;
+}
```
Comment (a-ilango) [inline prov/verbs/src/verbs_mr.c]: It seems fi_ibv_mr_ofi2ibv_access function is independent from enabling MR cache. Can you please move it into a separate patch?

File: prov/verbs/src/verbs_mr.c [hunk 7]
```diff
@@ -130,37 +154,292 @@ fi_ibv_mr_reg(struct fid *fid, const void *buf, size_t len,
...
+		return -errno;
+	return FI_SUCCESS;
+}
+
+int fi_ibv_mr_direct_reg(struct fi_ibv_domain *domain, void *buf,
+			 size_t len, uint64_t access,
+			 struct fi_ibv_mem_desc *md)
+{
+	return fi_ibv_mr_internal_reg(domain, buf, len, access, md);
+}
```
Comment (a-ilango) [inline prov/verbs/src/verbs_mr.c]: Just curious why this extra function is needed. It does nothing but call fi_ibv_mr_internal_reg.

File: prov/verbs/src/verbs_mr.c (hunk 4 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_mr.c]: yes, sure. the final solution will do that

File: prov/verbs/src/verbs_mr.c (hunk 6 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_mr.c]: Sure, we can. Done

**Comment by gladkovdmitry17** [comment]
HW: Intel(R) Xeon(R) CPU E5-2699 v4 @ 2.20GHz + Mellanox Connect X-4
The MPI implementation and both verbs/RDM and RxM are out of the box (only MR caching is enabled). The verbs provider is used from this patch.
MPI over OFI/RxM/verbs:
```
#---------------------------------------------------
# Benchmarking PingPong
# #processes = 2
#---------------------------------------------------
       #bytes #repetitions      t[usec]   Mbytes/sec
            0         1000         1.97         0.00
            1         1000         1.98         0.51
            2         1000         1.98         1.01
            4         1000         1.98         2.02
            8         1000         1.98         4.05
           16         1000         1.99         8.06
           32         1000         1.98        16.13
           64         1000         2.01        31.83
          128         1000         2.04        62.64
          256         1000         2.10       122.08
          512         1000         2.20       232.56
         1024         1000         2.39       427.91
         2048         1000         2.81       728.30
         4096         1000         3.74      1096.35
         8192         1000         5.31      1543.22
        16384         1000         8.77      1868.39
        32768         1000        10.89      3008.31
        65536          640        13.58      4825.46
       131072          320        19.58      6695.79
       262144          160        30.40      8624.04
       524288           80        51.36     10208.73
      1048576           40        94.64     11079.95
      2097152           20       181.07     11581.81
      4194304           10       353.79     11855.37

```

MPI over verbs/RDM:
```
#---------------------------------------------------
# Benchmarking PingPong
# #processes = 2
#---------------------------------------------------
       #bytes #repetitions      t[usec]   Mbytes/sec
            0         1000         1.17         0.00
            1         1000         1.18         0.84
            2         1000         1.18         1.69
            4         1000         1.18         3.40
            8         1000         1.18         6.76
           16         1000         1.21        13.26
           32         1000         1.24        25.82
           64         1000         1.25        51.14
          128         1000         1.30        98.76
          256         1000         1.75       146.29
          512         1000         2.21       231.31
         1024         1000         2.00       512.13
         2048         1000         2.40       852.47
         4096         1000         3.26      1255.29
         8192         1000         4.83      1696.94
        16384         1000         6.65      2465.05
        32768         1000         8.66      3784.48
        65536          640        11.54      5679.85
       131072          320        17.42      7526.07
       262144          160        27.52      9526.02
       524288           80        49.25     10645.48
      1048576           40        92.51     11334.44
      2097152           20       178.87     11724.22
      4194304           10       351.50     11932.57
```

I've added memory hooks to the verbs provider for now.
I'm going to separate this when the right architectural solution will be designed.

File: prov/util/src/util_mem_monitor.c [hunk 8]
```diff
@@ -93,14 +93,28 @@ int ofi_monitor_subscribe(struct ofi_notification_queue *nq,
...
+	       addr, len, subscription);
 	subscription->nq->monitor->unsubscribe(subscription->nq->monitor,
 						addr, len, subscription);
 	fastlock_acquire(&subscription->nq->lock);
 	dlist_remove_init(&subscription->entry);
 	subscription->nq->refcnt--;
+	while (dlist_remove_first_match(&subscription->nq->list,
+					util_monitor_find_match_subscription,
+					subscription))
+		;
```
Comment (shefty) [inline prov/util/src/util_mem_monitor.c]: Why is this search needed?  Shouldn't the dlist_remove_init() call above remove it?

Comment (gladkovdmitry17) [reply prov/util/src/util_mem_monitor.c]: Yes, this works. Revert this changes, because it just hides verbs problem
The problem was in verbs provider's code. Fixed and replaced by `dlist_remove_init()`

Thanks!

**Comment by gladkovdmitry17** [comment]
this patch produces warnings like this:
```
prov/verbs/src/verbs_domain.c(122): warning #1478: variable "__free_hook" (declared at line 172 of "/usr/include/malloc.h") was declared deprecated
        __free_hook = global_domain->prev_free_hook;
```

@shefty  Can we allow use deprecated variable via setting an appropriate compiler's option?

**Comment by shefty** [comment]
I'd be okay with disabling the warning, but I'd like to disable it within the smallest scope possible.  Is there an alternative that we can use instead?  This call was deprecated since it isn't thread safe.

**Comment by gladkovdmitry17** [comment]
> Is there an alternative that we can use instead? This call was deprecated since it isn't thread safe.

Yes, we can. `free`, `realloc` and other are defined as a week symbol in glibc. It means that they can be overwritten by our functions.
Jemalloc does this  - https://github.com/jemalloc/jemalloc/blob/7f1b02e3fa9de7e0bb5e2562994b5ab3b82c0ec3/src/jemalloc.c#L2494

**Comment by shefty** [comment]
I know that trying to hook malloc/free calls has always been problematic and error prone in practice.  I don't think we want libfabric to export these sort of calls.  My guess is that the broader community will be against it.  That said, I think there are situations where hooking the calls will work, and I want to enable those cases.

We need to brainstorm on how to best approach this.  Maybe it's time to look at how libfabric can make use of plug-ins that aren't just providers.

**Comment by gladkovdmitry17** [comment]
> Maybe it's time to look at how libfabric can make use of plug-ins that aren't just providers.

Do you mean to implement a kind of framework (separate library) that can be used by libfabric library?

**Comment by gladkovdmitry17** [comment]
> We need to brainstorm on how to best approach this.

@shefty I agree that we should do this, but can we go ahead with current solution in verbs provider (some places will be refactored a bit) to meet expectations of application that uses libfabric underneath.

When we found out the best approach how plug-ins should be exposed by libfabric and implement the Memory notifier plug-in (if I correctly got your idea), we will remove this from the verbs provider.
Btw, I'm fine to drive and put efforts to brainstorm, design and implement those ideas.

**Comment by shefty** [comment]
That's what I meant, but I'm not sure it's the right solution.  If we want to hook malloc/free, I definitely prefer that be done by some other library.  But for that to be useful, we need that library to somehow hook into the monitor framework.  That makes me wonder if we wouldn't find other uses for searching for and using external functionality -- much like we support external providers, but that's not an actual provider.

**Comment by shefty** [comment]
I agree with proceeding with using __free_hook for now.  I'd just like to disable the warning message around it, and re-enable the warning after, probably using some sort of pragma's.

**Comment by gladkovdmitry17** [comment]
> I agree with proceeding with using __free_hook for now.

Thanks!


> I'd just like to disable the warning message around it, and re-enable the warning after, probably using some sort of pragma's.

Ok, I'll propose the solution applicable for all compilers.

**Comment by gladkovdmitry17** [comment]
@shefty @a-ilango after applying several fixes and simplifications, the MR caching works well for the verbs/RDM and RxM/verbs/MSG providers.
Could you take a look, please?

You might be wondering why dlist_entry_... was introduced
The previous implementation uses dlist_init() for entries, it sets prev and next to an addr of entry itself. call dlist_empty() for a list checks that `head->next != head` is. If some entry was inserted and removed then, it leads to `head->next == removing entry`

**Comment by gladkovdmitry17** [comment]
```
FI_VERBS_MR_CACHE_ENABLE=1 date; ./runfabtests.sh -p ./ -c 192.168.65.3 -s 192.168.65.3 verbs | tee test.out
Fri Dec 22 17:36:47 MSK 2017
# Test                                                  Result
# --------------------------------------------------------------
fi_getinfo_test -p "verbs":                             Pass
fi_av_test -g 192.168.10.1 -n 1 -s 192.168.65.3 -p "verbs":      Pass
fi_dom_test -n 2 -p "verbs":                            Pass
fi_eq_test -p "verbs":                                  Pass
fi_cq_test -p "verbs":                                  Pass
fi_mr_test -p "verbs":                                  Pass
fi_cntr_test -p "verbs":                                Pass
fi_dgram g00n13s -p "verbs":                            Pass
fi_rdm g00n13s -p "verbs":                              Pass
fi_msg g00n13s -p "verbs":                              Pass
fi_cm_data -p "verbs":                                  Pass
fi_cq_data -p "verbs":                                  Pass
fi_dgram -p "verbs":                                  Notrun
fi_dgram_waitset -p "verbs":                          Notrun
fi_msg -p "verbs":                                      Pass
fi_msg_epoll -p "verbs":                                Pass
fi_msg_sockets -p "verbs":                              Pass
fi_poll -t queue -p "verbs":                          Notrun
fi_poll -t counter -p "verbs":                        Notrun
fi_rdm -p "verbs":                                      Pass
fi_rdm_rma_simple -p "verbs":                         Notrun
fi_rdm_rma_trigger -p "verbs":                        Notrun
fi_shared_ctx -p "verbs":                             Notrun
fi_shared_ctx --no-tx-shared-ctx -p "verbs":          Notrun
fi_shared_ctx --no-rx-shared-ctx -p "verbs":          Notrun
fi_shared_ctx -e msg -p "verbs":                      Notrun
fi_shared_ctx -e msg --no-tx-shared-ctx -p "verbs":      Pass
fi_shared_ctx -e msg --no-rx-shared-ctx -p "verbs":    Notrun
fi_shared_ctx -e dgram -p "verbs":                    Notrun
fi_shared_ctx -e dgram --no-tx-shared-ctx -p "verbs":    Notrun
fi_shared_ctx -e dgram --no-rx-shared-ctx -p "verbs":    Notrun
fi_rdm_tagged_peek -p "verbs":                          Pass
fi_scalable_ep -p "verbs":                            Notrun
fi_cmatose -p "verbs":                                  Pass
fi_rdm_shared_av -p "verbs":                          Notrun
fi_multi_mr -e msg -V -p "verbs":                     Notrun
fi_multi_mr -e rdm -V -p "verbs":                     Notrun
fi_recv_cancel -e rdm -V -p "verbs":                  Notrun
fi_unexpected_msg -e msg -i 10 -p "verbs":            Notrun
fi_unexpected_msg -e rdm -i 10 -p "verbs":            Notrun
fi_unexpected_msg -e dgram -i 10 -p "verbs":          Notrun
fi_unexpected_msg -e msg -S -i 10 -p "verbs":         Notrun
fi_unexpected_msg -e rdm -S -i 10 -p "verbs":         Notrun
fi_unexpected_msg -e dgram -S -i 10 -p "verbs":       Notrun
fi_msg_pingpong -I 5 -p "verbs":                        Pass
fi_msg_pingpong -I 5 -v -p "verbs":                     Pass
fi_msg_bw -I 5 -p "verbs":                              Pass
fi_msg_bw -I 5 -v -p "verbs":                           Pass
fi_rma_bw -e msg -o write -I 5 -p "verbs":              Pass
fi_rma_bw -e msg -o read -I 5 -p "verbs":               Pass
fi_rma_bw -e msg -o writedata -I 5 -p "verbs":          Pass
fi_rma_bw -e rdm -o write -I 5 -p "verbs":              Pass
fi_rma_bw -e rdm -o read -I 5 -p "verbs":               Pass
fi_rma_bw -e rdm -o writedata -I 5 -p "verbs":          Fail
fi_msg_rma -o write -I 5 -p "verbs":                    Pass
fi_msg_rma -o read -I 5 -p "verbs":                     Pass
fi_msg_rma -o writedata -I 5 -p "verbs":                Pass
fi_msg_stream -I 5 -p "verbs":                          Pass
fi_rdm_atomic -I 5 -o all -p "verbs":                 Notrun
fi_rdm_cntr_pingpong -I 5 -p "verbs":                 Notrun
fi_rdm_multi_recv -I 5 -p "verbs":                      Pass
fi_rdm_pingpong -I 5 -p "verbs":                        Pass
fi_rdm_pingpong -I 5 -v -p "verbs":                     Pass
fi_rdm_rma -o write -I 5 -p "verbs":                    Pass
fi_rdm_rma -o read -I 5 -p "verbs":                     Pass
fi_rdm_rma -o writedata -I 5 -p "verbs":                Pass
fi_rdm_tagged_pingpong -I 5 -p "verbs":                 Pass
fi_rdm_tagged_pingpong -I 5 -v -p "verbs":              Pass
fi_rdm_tagged_bw -I 5 -p "verbs":                       Pass
fi_rdm_tagged_bw -I 5 -v -p "verbs":                    Pass
fi_dgram_pingpong -I 5 -p "verbs":                    Notrun
fi_rc_pingpong -n 5 -p "verbs":                         Pass
fi_rc_pingpong -n 5 -e -p "verbs":                      Pass
# --------------------------------------------------------------
# Total Pass                                                44
# Total Notrun                                              28
# Total Fail                                                 1
# Percentage of Pass                                        97
# --------------------------------------------------------------
```

File: prov/verbs/src/verbs_mr.c (hunk 6 above)
Comment (a-ilango) [reply prov/verbs/src/verbs_mr.c]: It seems this hasn't been moved to a separate patch.

**Comment by a-ilango** [comment]
>You might be wondering why dlist_entry_... was introduced
>The previous implementation uses dlist_init() for entries, it sets prev and next to an addr of entry itself. call dlist_empty() for a list checks that head->next != head is. If some entry was inserted and removed then, it leads to head->next == removing entry

Not sure if I'm following this. If there's just one entry in the list, dlist_remove should correctly set head->next = head.

File: prov/verbs/src/fi_verbs.h [hunk 9]
```diff
@@ -334,29 +337,35 @@ typedef int(*fi_ibv_mr_reg_cb)(struct fi_ibv_domain *domain, void *buf,
...
-	struct fi_info		*info;
+	enum fi_ep_type			ep_type;
+	struct fi_ibv_rdm_cm		*rdm_cm;
+	struct slist			ep_list;
+	struct fi_info			*info;
 	/* This EQ is utilized by verbs/RDM and verbs/DGRAM */
-	struct fi_ibv_eq	*eq;
-	uint64_t		eq_flags;
+	struct fi_ibv_eq		*eq;
+	uint64_t			eq_flags;
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.h]: Please please avoid fixing unrelated white space issues in the future. It makes cherry-picking difficult. :)

File: prov/verbs/src/verbs_mr.c (hunk 6 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_mr.c]: > It seems this hasn't been moved to a separate patch.

Sorry, it has been missed during force push from auxiliary repo.

Now it should be there.

File: prov/verbs/src/fi_verbs.h (hunk 9 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/fi_verbs.h]: Sure, will take it into account in the future.

**Comment by gladkovdmitry17** [comment]
> Not sure if I'm following this. If there's just one entry in the list, dlist_remove should correctly set head->next = head.

Sorry, I had wrong thoughts regarding this. I've removed this

**Comment by gladkovdmitry17** [comment]
@shefty @a-ilango I guess, the patch is in a good shape now

Thanks for the comments and helping me out!

**Comment by a-ilango** [comment]
👍

**Comment by gladkovdmitry17** [comment]
> 👍

@shefty @a-ilango Since this patch doesn't affect default functionality of verbs provider and/or common code and other providers, I'm merging this patch.

To be able use this feature, user should specify `FI_VERBS_MR_CACHE_ENABLE=1`
